File: src/gx/commands/switch.py
"""gx switch — Fuzzy-find branch switcher with rich context."""

from __future__ import annotations

import typer

from gx.utils.display import console, print_error, print_info, print_success, print_warning
from gx.utils.git import (
    GitError,
    ensure_git_repo,
    get_current_branch,
    get_head_branch,
    is_clean_working_tree,
    run_git,
    time_ago,
)


def _get_branches_with_info() -> list[dict]:
    """Get all local branches with metadata."""
    fmt = "%(refname:short)\t%(authordate:iso8601)\t%(authorname)\t%(objectname:short)"
    output = run_git(["branch", f"--format={fmt}", "--sort=-committerdate"])
    if not output:
        return []

    branches = []
    try:
        current = get_current_branch()
    except GitError:
        current = ""

    for line in output.splitlines():
        parts = line.split("\t", 3)
        if len(parts) < 4:
            continue
        branches.append({
            "name": parts[0],
            "date": parts[1],
            "author": parts[2],
            "hash": parts[3],
            "current": parts[0] == current,
        })
    return branches


def switch(
    shortcut: str = typer.Argument(None, help='Use "-" to switch to previous branch.'),
) -> None:
    """Fuzzy branch switcher with rich context."""
    try:
        ensure_git_repo()
    except GitError as e:
        print_error(str(e))
        raise typer.Exit(1)

    # Handle gx switch -
    if shortcut == "-":
        try:
            run_git(["switch", "-"])
            new_branch = get_current_branch()
            print_success(f"Switched to {new_branch}")
        except GitError as e:
            if "no previous branch" in str(e).lower() or "invalid reference" in str(e).lower():
                print_error("No previous branch to switch to.")
            else:
                print_error(f"Failed to switch: {e}")
            raise typer.Exit(1)
        return

    # Warn about dirty working tree
    if not is_clean_working_tree():
        print_warning("You have uncommitted changes. They may conflict with the target branch.")

    branches = _get_branches_with_info()
    if not branches:
        print_info("No branches to switch to.")
        raise typer.Exit(0)

    # Filter out current branch
    try:
        current = get_current_branch()
    except GitError:
        current = ""

    selectable = [b for b in branches if b["name"] != current]
    if not selectable:
        print_info("No other branches to switch to.")
        raise typer.Exit(0)

    if len(selectable) == 1:
        target = selectable[0]["name"]
        try:
            run_git(["switch", target])
            print_success(f"Switched to {target}")
        except GitError as e:
            print_error(f"Failed to switch: {e}")
            raise typer.Exit(1)
        return

    # Launch Textual TUI
    try:
        from gx.commands._switch_tui import run_switch_tui
        selected = run_switch_tui(selectable, get_head_branch())
    except ImportError:
        # Fallback to simple selection if Textual is not available or TUI fails
        selected = _fallback_picker(selectable)

    if selected is None:
        print_info("Cancelled.")
        raise typer.Exit(0)

    try:
        run_git(["switch", selected])
        print_success(f"Switched to {selected}")
    except GitError as e:
        print_error(f"Failed to switch: {e}")
        raise typer.Exit(1)


def _fallback_picker(branches: list[dict]) -> str | None:
    """Simple numbered list picker as fallback."""
    console.print()
    console.print("[bold]Select a branch:[/bold]")
    console.print()
    for i, b in enumerate(branches, 1):
        age = time_ago(b["date"])
        console.print(f"  {i:>3}  {b['name']:<40} {age:<15} {b['author']}")

    console.print()
    try:
        choice = console.input("Enter number (or q to cancel): ")
        if choice.strip().lower() in ("q", ""):
            return None
        idx = int(choice.strip()) - 1
        if 0 <= idx < len(branches):
            return branches[idx]["name"]
        return None
    except (ValueError, EOFError, KeyboardInterrupt):
        return None

File: cmd/switch.go
package cmd

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/mubbie/gx-cli/internal/git"
	"github.com/mubbie/gx-cli/internal/ui"
	"github.com/spf13/cobra"
)

func init() {
	rootCmd.AddCommand(&cobra.Command{
		Use:   "switch [- | branch]",
		Short: "Branch switcher with search and rich context",
		Args:  cobra.MaximumNArgs(1),
		RunE:  runSwitch,
	})
}

type branchInfo struct {
	name    string
	date    string
	author  string
	current bool
}

func runSwitch(cmd *cobra.Command, args []string) error {
	if err := git.EnsureRepo(); err != nil {
		ui.PrintError(err.Error())
		return nil
	}

	// gx switch -
	if len(args) == 1 && args[0] == "-" {
		if _, err := git.Run("switch", "-"); err != nil {
			if strings.Contains(err.Error(), "no previous") || strings.Contains(err.Error(), "invalid") {
				ui.PrintError("No previous branch to switch to.")
			} else {
				ui.PrintError(fmt.Sprintf("Failed to switch: %s", err))
			}
			return nil
		}
		current, _ := git.CurrentBranch()
		ui.PrintSuccess(fmt.Sprintf("Switched to %s", current))
		return nil
	}

	if !git.IsClean() {
		ui.PrintWarning("You have uncommitted changes. They may conflict with the target branch.")
	}

	branches := getBranches()
	if len(branches) == 0 {
		ui.PrintInfo("No branches to switch to.")
		return nil
	}

	current, _ := git.CurrentBranch()
	var selectable []branchInfo
	for _, b := range branches {
		if b.name != current {
			selectable = append(selectable, b)
		}
	}

	if len(selectable) == 0 {
		ui.PrintInfo("No other branches to switch to.")
		return nil
	}

	if len(selectable) == 1 {
		if _, err := git.Run("switch", selectable[0].name); err != nil {
			ui.PrintError(fmt.Sprintf("Failed to switch: %s", err))
			return nil
		}
		ui.PrintSuccess(fmt.Sprintf("Switched to %s", selectable[0].name))
		return nil
	}

	selected := pickBranch(selectable)
	if selected == "" {
		ui.PrintInfo("Cancelled.")
		return nil
	}

	if _, err := git.Run("switch", selected); err != nil {
		ui.PrintError(fmt.Sprintf("Failed to switch: %s", err))
		return nil
	}
	ui.PrintSuccess(fmt.Sprintf("Switched to %s", selected))
	return nil
}

func getBranches() []branchInfo {
	out := git.RunUnchecked("branch", "--format=%(refname:short)\t%(authordate:iso8601)\t%(authorname)", "--sort=-committerdate")
	current, _ := git.CurrentBranch()
	var branches []branchInfo
	for _, line := range strings.Split(out, "\n") {
		parts := strings.SplitN(line, "\t", 3)
		if len(parts) < 3 || parts[0] == "" {
			continue
		}
		branches = append(branches, branchInfo{
			name:    parts[0],
			date:    parts[1],
			author:  parts[2],
			current: parts[0] == current,
		})
	}
	return branches
}

func pickBranch(branches []branchInfo) string {
	query := ""
	reader := newLineReader()

	for {
		var filtered []branchInfo
		if query == "" {
			filtered = branches
		} else {
			for _, b := range branches {
				if strings.Contains(strings.ToLower(b.name), query) {
					filtered = append(filtered, b)
				}
			}
		}

		fmt.Println()
		if len(filtered) == 0 {
			fmt.Println(ui.DimStyle.Render("  No branches match your search."))
		} else {
			for i, b := range filtered {
				age := git.TimeAgo(b.date)
				fmt.Printf("  %3d  %-40s %-15s %s\n", i+1, b.name, age, b.author)
			}
		}

		fmt.Println()
		if len(filtered) > 0 {
			fmt.Println(ui.DimStyle.Render("Enter a number to switch, text to filter, q to cancel"))
		} else {
			fmt.Println(ui.DimStyle.Render("Enter text to filter, q to cancel"))
		}

		choice := reader.read()
		if choice == "" || strings.ToLower(choice) == "q" {
			return ""
		}

		// Try as number
		if n, err := strconv.Atoi(choice); err == nil {
			idx := n - 1
			if idx >= 0 && idx < len(filtered) {
				return filtered[idx].name
			}
			fmt.Printf("  %s\n", ui.ErrorStyle.Render(fmt.Sprintf("Invalid number. Pick 1-%d.", len(filtered))))
			continue
		}

		query = strings.ToLower(choice)
	}
}

type lineReader struct{}

func newLineReader() lineReader { return lineReader{} }

func (lineReader) read() string {
	fmt.Print("> ")
	var line string
	fmt.Scanln(&line)
	return strings.TrimSpace(line)
}
